Python/ceval.c

```diff
@@ -2095,8 +2095,15 @@ _PyEval_ExceptionGroupMatch(PyObject* exc_value, PyObject *match_type,
         if (pair == NULL) {
             return -1;
         }
-        assert(PyTuple_CheckExact(pair));
-        assert(PyTuple_GET_SIZE(pair) == 2);
+
+        if (!PyTuple_CheckExact(pair) || PyTuple_Size(pair) != 2) {
+            PyErr_Format(PyExc_TypeError,
+                        "%s.split must return a 2-tuple",
+                        Py_TYPE(exc_value)->tp_name);
+            Py_DECREF(pair);
+            return -1;
+        }
+
         *match = Py_NewRef(PyTuple_GET_ITEM(pair, 0));
         *rest = Py_NewRef(PyTuple_GET_ITEM(pair, 1));
         Py_DECREF(pair);
```

Comment:
If `split()` returns a tuple of length more than 2, then it used to work but now it won't. So I think we need to change the check from `PyTuple_Size(pair) != 2` to `PyTuple_Size(pair) < 2` to make sure we don't break anything.

Reply:
This also means we need an entry in `Doc/whatsnew/3.14.rst`, in case there's some other edge case that used to work and now will not.

Reply:
I think we keep the strict len == 2 check and make the error messages more descriptive to make it easier to resolve if there's anyone out there who has actually made this mistake unintentionally

Reply:
We don't break things that used to work without deprecation. We also can't backport to stable versions something that could break working code.

This PR is to fix the crash. Tightening the API is a different problem.